doc/developer/design/20230225_compute_introspection_schema.md

```diff
@@ -0,0 +1,212 @@
+the case for each relation.
+
+Furthermore, sometimes the same logical value is reported by two
+relations, one reporting `clusterd`'s view of the world, and the other
+reporting `environmentd`'s. For example, `mz_compute_frontiers`
+vs. `mz_cluster_replica_frontiers`. However, the names do not make it
+clear what the distinction is. 
+
+#### Redundant `mz_` in names.
+
+Every relation's name begins with `mz_`. This is redundant, because
+anyone querying these relations probably already understand that they
+are using Materialize. Even if not, they are already in a namespace
+called `mz_internal`.
+
+### Suggested fixes for schema inconsistencies
+
+* All per-worker relations should have a corresponding "global" view
+  where the results are aggregated across workers in the way most
+  likely to be useful (for example, summing, taking the minimum, or
+  restricting to worker 0). Because these global views are more likely
+  to be useful than the underlying sources, they should have the basic
+  name (e.g., `mz_arrangement_sharing`). The underlying source should
+  have a name with `per_worker` inserted (e.g.,
+  `mz_per_worker_arrangement_sharing`). Furthermore, the per-worker
+  variant should be clearly less prominent in documentation (either at
+  the bottom of the page, or in a "see more" section that has to be
+  clicked to expand).
+* The expectations about per-worker relations should be clearly
+  documented (e.g., we should clearly say whether the values for each
+  worker are expected to be the same during normal operation).
+* All relations that report a simple count or sum should be views based on
+  an `_internal` variant that maintains the count or sum in the
+  Differential diff field. We may consider renaming this, since it
+  conflicts confusingly with the overall name `mz_internal`.
+* `raw` variants should be removed, and the underlying dataflows
+  should be in terms of intervals.
+* All histogram views or sources should have a per-bucket `sum`
+  column.
+* All per-replica sources and views should have their names prepended
+  with some string (perhaps `clusterd_`, though I'm open to
+  suggestions) that makes it obvious which ones they are. Furthermore,
+  the limitations of per-replica relations should be clearly
+  documented.
+* The `mz_` prefix should be removed from all names.
+
+### Possibly useful missing views
+
+#### Channel-to-operator mapping
+
+See [GitHub
+PR](https://github.com/MaterializeInc/materialize/pull/17825).
+
+#### Scope child to parent mapping
+
+The only way to understand the structure of scope nesting currently is
+to use the `mz_dataflow_addresses` relation and do some tedious list
+manipulation. A simple view of `(id, parent_id)` would make this much
+simpler.
+
+#### More views enriched with operator names, dataflow names, etc.
+
+(Not quite sure about this -- it's tricky to balance ease-of-use with
+not blowing up the set of documented relations)
+
```

Comment:
Are we up for rethinking the contents and structure of existing introspection sources? For example, I've been wondering if we really need to expose dataflow/channel addresses. If we are only interested in scopes, we could expose those more directly instead. And instead of exposing channels with their addresses, which then require some work to map them to the operators they connect, we could maybe expose channels with their operator IDs directly.

I think it would be a good idea to write down all the per-replica concepts we are interested in, their attributes and relations (e.g., as an ER diagram) and then translating that into introspection sources more or less directly. I appreciate that this is probably out-of-scope of this design doc.

Reply:
We can certainly discuss this here. I think the main things we care about are:

* Channels (and the number/size of messages sent on them)
* Operators (and the number/size of objects they retain, which is only relevant for arrangements)
* Scopes and their parent/child relationships
* Dataflows and the dependencies among them
* Frontiers

I agree that I can't immediately think of a good reason to expose addresses.

Reply:
Here is what I've come up with as a first draft:
![Blank diagram](https://user-images.githubusercontent.com/4521314/221614660-98bd9407-2bf9-4e29-80fb-5110c899af1e.png)

It's missing a couple thing (parks, delays), but most of what we currently have is there. The things marked with * are changing as part of dataflow processing. We should probably separate these from things that only change as part of dataflow creation/dropping, to minimize the size of changes. I think if you do that you end up somewhat close to what we already have, which is reassuring.

The most debatable thing about the approach I depicted is probably that grouping operators inside scopes requires you to use `WITH MUTUALLY RECURSIVE` to enumerate all the operators inside one dataflow, which will be great in half a year, but not today. We should probably connect the operators to their dataflow directly too.

Also, our current introspection sources mirror the structure of timely's/DD's logging quite closely. Having this more user-centric view will probably make the logging logic in clusterd more of a hassle to implement.

Reply:
Can you summarize the main differences between this proposal and what we have now?

Reply:
My intuition is that this will end up being a rather large change; we should probably discuss it as separate from this design doc (as you suggested initially)